File: C++/Level.3/단어변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

string s_word;
string t_word;
int n;
vector<string> lists;
unordered_map<string, bool> visited;

bool areTheyDifferOne(string &a, string &b) {
	int string_len = a.length();
    int cnt = 0;
    for(int i = 0; i < string_len; ++i) {
        if (a[i] != b[i]) {
            ++cnt;
        }
    }
    if (cnt != 1) {
        return false;
    }
    return true;
}

int BFSUtil(string s_word) {
	queue<pair<string, int>> q;
    q.push({s_word, 0});
    while (!q.empty()) {
      	pair<string, int> tmp = q.front();
        if (tmp.first == t_word) {
            return tmp.second;
        }
        visited[tmp.first] = 1;
        q.pop();
        for(int i = 0; i < n; ++i) {
            if (
                !visited[lists[i]] &&
                areTheyDifferOne(lists[i], tmp.first) 
               ) {
                q.push({lists[i], tmp.second + 1});
            }
        } 
    }
    return 0;
}

int BFS(string begin, string target, vector<string> words) {
   	s_word = begin;
    t_word = target;
    lists = words;
    n = words.size();
    return BFSUtil(begin);
}

int solution(string begin, string target, vector<string> words) {
	return BFS(begin, target, words);
}

File: JavaScript/Level.3/단어변환.js
function BFS(begin, target, words) {
    const n = words.length;

    function DifferOne(a, b) {
        let len = a.length;
        let cnt = 0;
        for (let i = 0; i < len; ++i) {
            if (a[i] !== b[i]) {
                ++cnt;
            }
        }
        if (cnt === 1) return true;
        return false;
    }

    function BFSUtil(begin) {
        let queue = [];
        queue.push([begin, 0]);
        while (queue.length > 0) {
            let word = queue[0][0],
                num = queue[0][1];
            if (word === target) {
                return num;
            }
            if (num > n) return 0;
            queue.shift();
            for (let i = 0; i < n; ++i) {
                if (DifferOne(word, words[i])) {
                    queue.push([words[i], num + 1]);
                }
            }
        }
        return 0;
    }
    return BFSUtil(begin);
}

function solution(begin, target, words) {
    return BFS(begin, target, words);
}
